cilium: pcap recorder agent management

borkmann opened the pull request:
See commit msg.

Agent management middle layer for BPF LB code: https://github.com/cilium/cilium/pull/15376

Follow-ups: https://github.com/cilium/cilium/issues/15712

Hubble API code that @gandro is working on which is on top of this PR: https://github.com/cilium/cilium/pull/15680 and https://github.com/cilium/hubble/pull/530

borkmann committed:
cilium, api: add agent configuration API for recorders

Add an agent API model to the recorder. Idea is that a recorder object
is defined by a unique ID, capture length and an array of one or more
wildcarded filters. The higher level entity that is configuring this API
(e.g. Hubble) has awareness of the IDs and reads the perf RB where it
can then output the content to different pcap files based on the ID.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: add auto-generated recorder API for client/server

Auto generated via `make generate-api` from api spec. Kept separate to
ease review on the actual api/v1/openapi.yaml changes.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, cli: add cilium recorder base command

Add initial `cilium recorder` cli which will be used in similar way as
the `cilium service` cli, that is, for introspection and management of
pcap recorder middle layer. This complements `cilium bpf recorder` in
that the former is for the agent middle layer while the latter is for
the actual BPF datapath map.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, cli: add cilium recorder list command

Add a `cilium recorder list` command to dump all existing recorders
from the agent.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, client: add recorder API to client package

Add an initial client recorder package which the cilium cli will
be using in upcoming commits.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, bugtool: add cilium recorder list to sysdump

For ease of introspection on user issues, add `cilium recorder list` to
the bug tool dump. We already have `cilium bpf recorder list` in there
from BPF side, thus this complements the agent view.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, doc: add cmdref documentation for cilium recorder

Add all auto-generated cmdref doc bits.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, agent: add recorder server facing API

Hook up stubs in the daemon for the auto-generated recorder API for
all /recorder/{id}/ and /recorder/ paths.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, recorder: add initial recorder package

Trivial start of recorder package which adds a Recorder struct that
the server-facing API handler will be using.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, recorder: add middle layer implementation

Add the core part of the recorder handling, that is, tracking of
recorder entries as well as mask generation and back'n'forth
conversion of the API model.

We expose the following APIs from the package for external users
to get, create/update, or delete a recorder:

  - func (r *Recorder) RetrieveRecorder(id ID) (*RecInfo, error)
  - func (r *Recorder) UpsertRecorder(recInfoNew *RecInfo) (bool, error)
  - func (r *Recorder) DeleteRecorder(id ID) (bool, error)

And one more API to get a list of all current recorders:

  - func (r *Recorder) RetrieveRecorderSet() []*RecInfo

A recorder object consists of a unique id, capture length and a
list of corresponding filters:

  type RecInfo struct {
	id      ID
	capLen  uint16
	filters []RecorderTuple
  }

A single filter is defined as:

  type RecorderTuple struct {
	srcPrefix net.IPNet
	srcPort   uint16
	dstPrefix net.IPNet
	dstPort   uint16
	proto     u8proto.U8proto
  }

The recorder id management is delegated to caller code currently, but
in future this could also be managed by the recorder package. Meaning,
Hubble will have an internal id generator plus an internal mapping of
id-to-pcap file. The id is not exposed to users, it is used as a means
to correlate incoming recorder perf RB events which the Hubble reader
will pick up and then push to the right pcap file.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: hook up to recorder middle layer

Complete the server API and extend stubs to integrate pkg/recorder bits,
so we are able to handle actual requests.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, cli: add cilium recorder get command

Add `cilium recorder get` command to the cilium cli to dump a single
recorder based on the given id.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, cli: add cilium recorder delete command

Add `cilium recorder delete` command to the cilium cli to delete a single
recorder based on the given id.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, cli: add cilium recorder update command

Add `cilium recorder update` command to the cilium cli to create or
update a single recorder based on the given id. Cilium cli usage:

  # cilium recorder update --id 2 --caplen 100 --filters="10.0.0.0/8 0 1.1.1.1/32 80 TCP, 1.1.1.1/32 80 10.0.0.0/8 0 TCP"

The above adds a recorder with id 2 and max capture length of 100 byte
which holds two wildcard filters to capture all traffic from/to subnet
10.0.0.0/8 to a service IP 1.1.1.1/32 with port 80 and protocol TCP.

The main purpose here is for manual testing. Idea is similar to `cilium
service`, that is, regular service management is done through K8s watchers
and the cli itself provides a means for manual editing/testing. The same
is with recorders where Hubble will be the main infra to manage them.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: add wildcard mask status API for recorders

Next up, extend the recorder API with a getter for all generated masks
from the user installed tuples. This is exposed through `/recorder/masks`
and helps for introspection purpose. The modelling is similar to the
earlier RecorderSpec.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: add auto-generated recorder mask API for client/server

Auto generated via `make generate-api` from api spec. Also, kept separate
to ease review on the actual api/v1/openapi.yaml changes.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, client: add recorder mask API to client package

Extend the client recorder package with a GetRecorderMasks() helper to
request retrieval of all masks.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, agent: add recorder mask server facing API

Hook up stubs in the daemon for the auto-generated recorder mask API
for the /recorder/masks path.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, api: hook up to recorder mask middle layer

Implement the getRecorderMaskList() stub from the agent, and extend the
recorder package with helpers to dump all masks and convert them to the
API model.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, cli: add masks to cilium recorder list command

Extend the `cilium recorder list` to also include all masks that the
agent is tracking.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, recorder: add priority to mask based on ones

Extend the masks and add a priority field to the API which allows for
the `cilium recorder list` cli command to sort and dump the masks in
the same order as the datapath would have.

Example dump:

  # ./cilium/cilium recorder list
  ID   Capture Length   Wildcard Filters
  10   <= 100           f00d::1/128:0      ->   f00e::1/128:0   ANY
                        5.5.5.5/32:0       ->   3.3.3.3/32:0    ANY

  Users   Priority      Wildcard Masks
  1       256           ffffffffffffffffffffffffffffffff:0   ->   ffffffffffffffffffffffffffffffff:0   0
  1       64            ffffffff:0                           ->   ffffffff:0                           0

The order is based on the number of ones, which is necessary given we
want to give longer prefixes a higher priority for the lookup.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, recorder: flush recorder bpf maps upon agent init

We currently don't support restore functionality from Hubble API side,
therefore flush the BPF maps upon agent restart instead of adding restore
functionality.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, recorder: upsert/delete recorder bpf map entries

Add helpers to the recorder package which craft CaptureWcard{4,6} structs
from tuples and push them down to the BPF datapath maps.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, recorder: expose RecorderTuple and RecInfo for Hubble

Given Hubble will be consuming the recorder API, expose all RecInfo and
RecorderTuple fields, so they can be populated from external.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

gandro commented on api/v1/openapi.yaml:
```diff
@@ -2418,6 +2502,98 @@ definitions:
+        description: Layer 4 protocol
+        type: string
+        enum:
+        - TCP
+        - UDP
+        - ANY
```
Can we also support ICMP/ICMPv6? The proposed Hubble API and the `u8proto` proto package already assume they work, but I'm less certain about the datapath.

borkmann replied:
For the load balancer itself, it won't be able to IPIP encap ICMP traffic. The LB currently only supports UDP+TCP. I think once we generalize the pcap recorder further, we could integrate it.

gandro commented on cilium/cmd/recorder_list.go:
```diff
@@ -0,0 +1,135 @@
+				filter.DstPrefix, filter.DstPort,
+				filter.Protocol)
+			fmt.Fprintln(w, str)
+		}
```
Do we need to check for `len(spec.Filters) > 0` here? I haven't seen any checks that stop a user from submitting a recording without filters.

borkmann replied:
Actually all the code can handle it, e.g. I could reserve an empty id/caplen and then add/del filter later on.

gandro commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,475 @@
+	mask  RecorderMask
+}
+
+type RecQueue struct {
```
nit: This type should be made private
```suggestion
type recQueue struct {
```

gandro commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,475 @@
+	Proto     u8proto.U8proto
+}
+
+type RecorderMask struct {
```
nit: This type maybe be made private if the fields of `RecMask` a intended to be private (which I believe the are)
```suggestion
type recorderMask struct {
```

aanm replied:
this does not seem to be resolved

gandro commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,475 @@
+}
+
+func queuePurge(q []*RecorderTuple, i int) []*RecorderTuple {
+	q[i] = q[len(q)-1]
```
nit: To allow for GC to actually clean up the old tuple, also set
```suggestion
	q[i] = q[len(q)-1]
	q[len(q)-1] = nil
```

borkmann replied:
ack, makes sense, pushed with setting `q[len(q)-1] = nil`.

gandro commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,475 @@
+}
+
+func (r *Recorder) createRecInfoLocked(ri *RecInfo, withID bool) bool {
+	if withID {
+		r.recByID[ri.ID] = ri
+	}
```
nit: `withID` is never false here in the current version. Do we still need it?

borkmann replied:
I had it for symmetry for both upsert/delete ops.

gandro reviewed (commented):
Nice! Looks good, I haven't spotted any bugs in my initial review - just a few minor nits.

Will do another review once it's out of draft mode.

borkmann committed:
cilium, loader: refactor XDP bits into separate function

Given we need to recompile and reload the BPF XDP datapath for new masks,
refactor and export ReinitializeXDP(), so it can be called from other
packages.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, loader: add extra args for XDP compilation

Extend the ReinitializeXDP() helper with an extraCArgs string slice so
that callers are able to pass in extra compilation flags. Also, add a
-Dcapture_enabled=0 to the initial compilation to make sure we initially
disable (and thus compile out) any recorder bits from XDP given upon
start there are no recorders anyway.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

borkmann committed:
cilium, recorder: rebuild upon wildcard mask change

Add initial implementation of the recorder packages triggerDatapathRegenerate()
handling. This will either disable or enable `capture_enabled` based on the
current masks that the agent tracks. Also, this will generate an ordered mask
list which is being pushed down to PREFIX_MASKS{4,6} for compilation. The
triggerDatapathRegenerate() helper will only be called when the mask set changes
but not otherwise, thus any subsequent API calls that do not need recompilation
won't incur this overhead.

Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>

brb commented on pkg/datapath/loader/base.go:
```diff
@@ -217,6 +217,16 @@ func (l *Loader) reinitializeIPSec(ctx context.Context) error {
 	return nil
 }
 
+func (l *Loader) ReinitializeXDP(ctx context.Context) error {
```
Nit: `ReinitializeXDP` reads as if the XDP prog has existed before which is not always the case. `CompileAndLoadXDP`?

borkmann replied:
I kept the existing naming conventions in this file, e.g. `Reinitialize()`, `reinitializeIPSec()`, etc.

brb commented on pkg/datapath/loader/base.go:
```diff
@@ -411,11 +421,8 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
-			log.WithError(err).Fatal("Failed to compile XDP program")
-		}
+	if err := l.ReinitializeXDP(ctx); err != nil {
+		log.WithError(err).Fatal("Failed to compile XDP program")
```
Nit: s/Failed to compile XDP program/Failed to compile or load XDP program/

borkmann replied:
ack, will change

brb commented on pkg/datapath/loader/base.go:
```diff
@@ -421,7 +421,8 @@ func (l *Loader) Reinitialize(ctx context.Context, o datapath.BaseProgramOwner,
 	defer cancel()
 
-	if err := l.ReinitializeXDP(ctx); err != nil {
+	extraArgs := []string{"-Dcapture_enabled=0"}
```
Can't we have in the XDP code instead?
```
#ifndef capture_enabled
#define capture_enabled 0
#endif
```

borkmann replied:
It's in the datapath as-is so it gets compile tested for XDP from normal make.

brb commented on pkg/recorder/recorder.go:
```diff
@@ -163,8 +168,93 @@ func (t *RecorderTuple) isIPv4() bool {
+	return bits == 32
+}
+
+func (m *RecorderMask) genMacroSpec() string {
```
As a follow up, we could use https://golang.org/pkg/text/template/ to make this a bit more readable.

brb commented on pkg/recorder/recorder.go:
```diff
@@ -163,8 +168,93 @@ func (t *RecorderTuple) isIPv4() bool {
+	return bits == 32
+}
+
+func (m *RecorderMask) genMacroSpec() string {
```
I think this code can be unit tested.

borkmann replied:
Yep, I'll follow up with unit test code given we need some overall anyway.

gandro commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,628 @@
+			string(drune[0:8]), string(drune[8:16]))
+		srune := []rune(m.srcMask.String())
+		spec += fmt.Sprintf(".saddr={.d1=0x%s,.d2=0x%s},",
+			string(srune[0:8]), string(srune[8:16]))
```
I don't believe this is correct. It yields `{.daddr={.d1=0xffffffff,.d2=0xffffffff}` for a mask of `ffffffffffffffffffffffff00000000` (/96).

```suggestion
		dstMask := m.dstMask.String()
		spec += fmt.Sprintf(".daddr={.d1=0x%s,.d2=0x%s},",
			dstMask[0:16], dstMask[16:32])
		srcMask := m.srcMask.String()
		spec += fmt.Sprintf(".saddr={.d1=0x%s,.d2=0x%s},",
			srcMask[0:16], srcMask[16:32])
```

borkmann replied:
Sigh, good catch, you are correct! I'll repush once net-next CI is through. Also, thx for the simplification!

brb commented on pkg/datapath/loader/base.go:
```diff
@@ -217,10 +217,10 @@ func (l *Loader) reinitializeIPSec(ctx context.Context) error {
 }
 
-func (l *Loader) ReinitializeXDP(ctx context.Context) error {
+func (l *Loader) ReinitializeXDP(ctx context.Context, extraArgs []string) error {
```
Nit: s/extraArgs/extraCArgs/ to denote the fact that they are used for compilation.

borkmann replied:
Ok, will do

gandro commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,625 @@
+			return nil, fmt.Errorf("Recorder source (%s) and destination (%s) prefix not supported by agent config",
+				f.SrcPrefix, f.DstPrefix)
+		}
+		port, err := strconv.Atoi(mf.DstPort)
```
nit: `strconv.ParseUint(mf.DstPort, 10,16)` would check if the port actually fits into a uint16

borkmann replied:
good idea, done, thx

brb commented on pkg/recorder/recorder.go:
```diff
@@ -163,8 +169,89 @@ func (t *RecorderTuple) isIPv4() bool {
+			extraArgs = append(extraArgs, masks6)
+		}
+	}
+	err := l.ReinitializeXDP(r.ctx, extraArgs)
```
Do we need to protect the `ReinitializeXDP` method with a lock? Could it be that a policy change (?) would trigger the datapath regeneration?

borkmann replied:
It's serialized implicitly here with the recorder lock. XDP doesn't have policy code in there yet, but if we add more regeneration contexts, we need to add a global XDP regen lock, agree.

brb replied:
I meant that https://github.com/cilium/cilium/blob/935d816a59ec9af92b41e9ba1f4d8412c66cc83b/pkg/datapath/loader/base.go#L436 is called by `(*Loader).Reinitialize(...)`. The latter can be called from multiple places (grep for `Loader().Reinitialize`).

gandro commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,625 @@
+			return nil, err
+		}
+		f.DstPort = uint16(port)
+		port, err = strconv.Atoi(mf.SrcPort)
```
nit: `strconv.ParseUint(mf.SrcPort, 10,16)` would check if the port actually fits into a uint16

borkmann replied:
ditto

brb reviewed (commented):
I did some nitpicking of the datapath regeneration related bits, assuming that Sebastian reviewed the cli, API, recorder bits.

My general observation that the code might lack comments (e.g. explaining the need for the queuing, or why we don't regenerate the datapath if the mask doesn't change, etc.). Lots of context is in the git commit messages which after code changes gets difficult to retrieve. Also, it would be good to get a review from someone who is not familiar with the idea.

Also, maybe some commits can be squashed together? E.g. `cilium, recorder: adapt RecorderMask exposure` or `cilium, doc: add cmdref documentation for cilium recorder`. It would make the reviewing easier.

gandro reviewed (approved):
Looks good to me, though I'm not a high-confidence reviewer when it comes to the datapath bits.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -2418,6 +2490,61 @@ definitions:
+        type: string
+      src-port:
+        description: Layer 4 source port, zero (or in future range)
+        type: string
```
no integer for ports?

borkmann replied:
Yeah, the string here is intentional given in future this could also be extended to port range, so wanted to keep this flexible.

aanm commented on api/v1/openapi.yaml:
```diff
@@ -2418,6 +2490,61 @@ definitions:
+        type: string
+      dst-port:
+        description: Layer 4 destination port, zero (or in future range)
+        type: string
```
no integer for ports?

borkmann replied:
same here

aanm commented on cilium/cmd/recorder_update.go:
```diff
@@ -0,0 +1,104 @@
+			fallthrough
+		case models.RecorderFilterProtocolUDP:
+			fallthrough
+		case models.RecorderFilterProtocolANY:
```
nit:
```golang
case models.RecorderFilterProtocolTCP,
    models.RecorderFilterProtocolUDP,
    models.RecorderFilterProtocolANY:
```

aanm commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,632 @@
+	return r.applyDatapath(triggerRegen)
+}
+
+func deepCopyPrefix(p net.IPNet) net.IPNet {
+	out := net.IPNet{
+		IP:   make([]byte, len(p.IP)),
+		Mask: make([]byte, len(p.Mask)),
+	}
+	copy(out.IP, p.IP)
+	copy(out.Mask, p.Mask)
+	return out
+}
+
+func deepCopyRecInfo(recInfo *RecInfo) *RecInfo {
+	ri := &RecInfo{
+		ID:      recInfo.ID,
+		CapLen:  recInfo.CapLen,
+		Filters: []RecorderTuple{},
+	}
+	for _, filter := range recInfo.Filters {
+		f := RecorderTuple{
+			SrcPort: filter.SrcPort,
+			DstPort: filter.DstPort,
+			Proto:   filter.Proto,
+		}
+		f.SrcPrefix = deepCopyPrefix(filter.SrcPrefix)
+		f.DstPrefix = deepCopyPrefix(filter.DstPrefix)
+		ri.Filters = append(ri.Filters, f)
+	}
+	return ri
+}
```
Please use the deepcopy generators that we have instead of doing it manually.

borkmann replied:
Where do I find these, do you have some pointers?

aanm replied:
here:
https://github.com/cilium/cilium/blob/6483724330b59947dce0c400660561f410031830/pkg/bpf/doc.go
https://github.com/cilium/cilium/blob/070214f089e78bf047ce671bd15d762174654ab6/pkg/bpf/endpoint.go#L34
https://github.com/cilium/cilium/blob/ae23400cb944c722f629b9b3f0d7e8af58af1b34/Makefile#L427

aanm commented on pkg/recorder/recorder.go:
```diff
@@ -0,0 +1,632 @@
+	return recList
+}
+
+func deepCopyMask(m net.IPMask) net.IPMask {
+	out := make([]byte, len(m))
+	copy(out, m)
+	return out
+}
+
+func deepCopyRecMask(recMask *RecMask) *RecMask {
+	rm := &RecMask{
+		users: recMask.users,
+		prio:  recMask.prio,
+		mask: recorderMask{
+			srcPort: recMask.mask.srcPort,
+			dstPort: recMask.mask.dstPort,
+			proto:   recMask.mask.proto,
+		},
+	}
+	rm.mask.srcMask = deepCopyMask(recMask.mask.srcMask)
+	rm.mask.dstMask = deepCopyMask(recMask.mask.dstMask)
+	return rm
+}
```
same here for the deepcopies

aanm reviewed (changes requested):
I got a little surprised to see that this feature only works in lb-mode :cry: ...

ti-mo commented on pkg/datapath/loader/base.go:
```diff
@@ -217,6 +217,16 @@ func (l *Loader) reinitializeIPSec(ctx context.Context) error {
 	return nil
 }
 
+func (l *Loader) ReinitializeXDP(ctx context.Context, extraCArgs []string) error {
```
- This method can be unexported, it's only used in `(*Loader).Reinitialize()` below.
- As discussed offline, a comment describing how and why to use this would be welcome. :+1: 
- Since this method only really takes strings, what about making it variadic?
```suggestion
func (l *Loader) ReinitializeXDP(ctx context.Context, extraCArgs ...string) error {
```
Then it can later be called as
```go
l.ReinitializeXDP(ctx, "-Dcapture_enabled=0")
```
instead of
```go
l.ReinitializeXDP(ctx, []string{"-Dcapture_enabled=0"})
```

borkmann replied:
Thanks! I added comments to exposed functions. The ReinitializeXDP() cannot be unexported given we use it from the recorder right now. Re variadic, no strong opinion from my side, it feels potentially a bit more complex given we pass in different number of extra args from the recorder side into a single ReinitializeXDP. I presume to make that work we would need to split the code have have two separate calls into ReinitializeXDP.. don't think it's really making it much better tbh.

ti-mo reviewed (approved):
Some nits on `pkg/datapath/loader/`, but looks great! :+1:

qmonnet reviewed (approved):
I didn't review thoroughly, but on a brief look everything seems in order.

I understand that the only change since last CI run (green) was the addition of a comment, and that the DeepCopy issue has been discussed offline with André. In that light I'm going ahead and merging the PR.
